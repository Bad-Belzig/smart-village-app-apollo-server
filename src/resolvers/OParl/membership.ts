import { FilterQuery } from 'mongoose';

import {
  IMembership,
  IMembershipSchema,
  Membership,
  Organization,
  Person,
} from '../../models';

export const membershipResolvers = {
  Query: {
    oParlMemberships: (
      _,
      args: {
        externalIds?: string[];
        keyword?: string[];
        offset?: number;
        pageSize?: number;
      },
    ) => {
      const filter: FilterQuery<IMembershipSchema> = {};

      if (args.keyword?.length) filter.keyword = { $all: args.keyword };

      if (args.externalIds) filter.externalId = { $in: args.externalIds };

      return Membership.find(filter);
    },
  },
  OParlMembership: {
    person: (parent: IMembership) =>
      Person.findOne({ externalId: parent.person }),
    organization: (parent: IMembership) =>
      Organization.findOne({ externalId: parent.organization }),
    onBehalfOf: (parent: IMembership) =>
      Organization.findOne({ externalId: parent.onBehalfOf }),
  },
};
