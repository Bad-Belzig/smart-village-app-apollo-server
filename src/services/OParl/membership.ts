import { fetch } from 'apollo-env';

import { Membership } from '../../models';
import { parseMembership } from '../../parser';

// import flow
// 1. fetch json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importMembership = async (url: string) => {
  try {
    const response = await fetch(url);

    if (response.ok) {
      const json = await response.json();

      // TODO: check for previous existence

      const lt = new Membership(parseMembership(json));

      await lt.validate();

      // TODO: check for person and organizations to import

      return lt.save();
    } else {
      throw new Error(
        `Error while fetching Membership from : ${url}! \n response.status: ${response.status}`,
      );
    }
  } catch (e) {
    console.log(e);
  }
};
