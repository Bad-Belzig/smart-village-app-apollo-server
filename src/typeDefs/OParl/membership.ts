import { gql } from 'apollo-server-core';

import { oParlBase } from './oParlBase';

export const membershipGQL = gql`
  extend type Query {
    oParlMemberships(
      externalIds: [String!]
      keyword: [String!]
      offset: Int
      pageSize: Int
    ): [OParlMembership!]
  }

  type OParlMembership {
    endDate: Date
    onBehalfOf: OParlOrganization
    organization: OParlOrganization
    person: OParlPerson
    role: String
    startDate: Date
    votingRight: Boolean
    ${oParlBase}
  }
`;
