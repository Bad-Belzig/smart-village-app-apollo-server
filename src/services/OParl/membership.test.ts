import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../testHelpers';
import { importMembership } from './membership';

const helperFuncs = setupTestDataBase('test_membership_import');

// TODO: change test url to something within our control
const testUrl =
  'https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/membership/1-2-115-13';

describe('importing a meeting', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importMembership, testUrl));

  it(
    'import returns undefined on faulty url',
    invalidUrlTest(importMembership),
  );

  afterAll(helperFuncs.afterAll);
});
