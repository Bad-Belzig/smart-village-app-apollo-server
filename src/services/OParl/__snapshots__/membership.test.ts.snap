// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a meeting imports properly 1`] = `
Object {
  "__v": 0,
  "created": 2017-01-30T17:23:35.000Z,
  "externalId": "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/membership/1-2-115-13",
  "keyword": Array [],
  "modified": 2017-09-22T05:31:08.000Z,
  "organization": "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/organization/2-115",
  "person": "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/person/13",
  "role": "Ratsfrau",
  "startDate": 1950-01-01T00:00:00.000Z,
  "type": "https://schema.oparl.org/1.0/Membership",
}
`;
